File: Backend/src/modules/food/dining/models/diningProfile.model.js
import mongoose from 'mongoose';

export const DINING_PROFILE_STATUSES = ['draft', 'pending', 'approved', 'rejected', 'suspended'];

const imageSchema = new mongoose.Schema(
    {
        url: { type: String, required: true, trim: true },
        publicId: { type: String, trim: true, default: '' },
        caption: { type: String, trim: true, default: '', maxlength: 120 }
    },
    { _id: false }
);

const statusHistorySchema = new mongoose.Schema(
    {
        status: { type: String, enum: DINING_PROFILE_STATUSES, required: true },
        note: { type: String, trim: true, default: '' },
        byRole: { type: String, trim: true, default: '' },
        byId: { type: mongoose.Schema.Types.ObjectId },
        at: { type: Date, default: Date.now }
    },
    { _id: false }
);

/**
 * One dining profile per restaurant. Created by the restaurant from its own panel
 * (never during onboarding) and reviewed by admin. Dining features unlock for the
 * restaurant only once `status === 'approved'`.
 */
const foodDiningProfileSchema = new mongoose.Schema(
    {
        restaurantId: {
            type: mongoose.Schema.Types.ObjectId,
            ref: 'FoodRestaurant',
            required: true,
            unique: true,
            index: true
        },
        status: { type: String, enum: DINING_PROFILE_STATUSES, default: 'pending', index: true },

        // Request details (shown to admin in the request list)
        categories: [{ type: mongoose.Schema.Types.ObjectId, ref: 'FoodDiningCategory', index: true }],
        about: { type: String, required: true, trim: true, maxlength: 2000 },
        cuisines: [{ type: String, trim: true, maxlength: 60 }],
        amenities: [{ type: String, trim: true, maxlength: 60 }],
        costForTwo: { type: Number, required: true, min: 0 },
        seatingCapacity: { type: Number, required: true, min: 1 },
        contactName: { type: String, required: true, trim: true, maxlength: 120 },
        contactPhone: { type: String, required: true, trim: true, maxlength: 20 },
        coverImage: { type: imageSchema, required: true },
        gallery: {
            type: [imageSchema],
            default: [],
            validate: {
                validator: (v) => Array.isArray(v) && v.length >= 1 && v.length <= 10,
                message: 'Between 1 and 10 gallery images are required'
            }
        },
        menuImages: { type: [imageSchema], default: [] },

        // Booking configuration (restaurant editable after approval)
        bookingWindowDays: { type: Number, default: 30, min: 1, max: 90 },
        slotDurationMins: { type: Number, default: 60, min: 15, max: 240 },
        maxGuestsPerBooking: { type: Number, default: 12, min: 1, max: 50 },
        minAdvanceMins: { type: Number, default: 30, min: 0, max: 1440 },
        autoConfirm: { type: Boolean, default: false },
        isOnline: { type: Boolean, default: true, index: true },

        // Denormalized for fast public listing/search
        restaurantName: { type: String, trim: true, default: '' },
        city: { type: String, trim: true, default: '', index: true },
        zoneId: { type: mongoose.Schema.Types.ObjectId, ref: 'FoodZone', index: true },
        location: {
            type: { type: String, enum: ['Point'], default: 'Point' },
            coordinates: { type: [Number], default: undefined }
        },
        ratingAvg: { type: Number, default: 0, min: 0, max: 5 },
        ratingCount: { type: Number, default: 0, min: 0 },
        totalBookings: { type: Number, default: 0, min: 0 },

        rejectionReason: { type: String, trim: true, default: '' },
        submittedAt: { type: Date, default: Date.now },
        reviewedAt: { type: Date },
        reviewedBy: { type: mongoose.Schema.Types.ObjectId, ref: 'User' },
        statusHistory: { type: [statusHistorySchema], default: [] }
    },
    {
        collection: 'food_dining_profiles',
        timestamps: true
    }
);

foodDiningProfileSchema.index({ status: 1, createdAt: -1 });
foodDiningProfileSchema.index({ status: 1, isOnline: 1, city: 1 });
foodDiningProfileSchema.index({ location: '2dsphere' }, { sparse: true });

export const FoodDiningProfile = mongoose.model(
    'FoodDiningProfile',
    foodDiningProfileSchema,
    'food_dining_profiles'
);

File: foodrestaurent/lib/features/dining/domain/dining_profile_model.dart
class DiningCategoryOption {
  final String id;
  final String name;
  final String image;

  const DiningCategoryOption({
    required this.id,
    required this.name,
    this.image = '',
  });

  factory DiningCategoryOption.fromJson(Map<String, dynamic> json) {
    return DiningCategoryOption(
      id: (json['id'] ?? json['_id'] ?? '').toString(),
      name: (json['name'] ?? '').toString(),
      image: json['image'] is Map
          ? (json['image']['url'] ?? '').toString()
          : (json['image'] ?? '').toString(),
    );
  }
}

class DiningProfileModel {
  final String id;
  final String restaurantId;
  final String restaurantName;
  final String status; // 'pending', 'approved', 'rejected', 'suspended'
  final String about;
  final List<DiningCategoryOption> categories;
  final List<String> cuisines;
  final List<String> amenities;
  final double costForTwo;
  final int seatingCapacity;
  final String contactName;
  final String contactPhone;
  final String coverImage;
  final List<String> gallery;
  final List<String> menuImages;
  final String rejectionReason;
  final bool isOnline;
  final bool autoConfirm;

  const DiningProfileModel({
    required this.id,
    required this.restaurantId,
    required this.restaurantName,
    required this.status,
    required this.about,
    required this.categories,
    required this.cuisines,
    required this.amenities,
    required this.costForTwo,
    required this.seatingCapacity,
    required this.contactName,
    required this.contactPhone,
    required this.coverImage,
    required this.gallery,
    required this.menuImages,
    required this.rejectionReason,
    required this.isOnline,
    required this.autoConfirm,
  });

  factory DiningProfileModel.fromJson(Map<String, dynamic> json) {
    final rawCats = json['categories'] as List? ?? [];
    return DiningProfileModel(
      id: json['id']?.toString() ?? json['_id']?.toString() ?? '',
      restaurantId: json['restaurantId']?.toString() ?? '',
      restaurantName: json['restaurantName']?.toString() ?? '',
      status: json['status']?.toString() ?? 'pending',
      about: json['about']?.toString() ?? '',
      categories: rawCats
          .map((c) => DiningCategoryOption.fromJson(Map<String, dynamic>.from(c as Map)))
          .toList(),
      cuisines: (json['cuisines'] as List? ?? []).map((e) => e.toString()).toList(),
      amenities: (json['amenities'] as List? ?? []).map((e) => e.toString()).toList(),
      costForTwo: (json['costForTwo'] as num?)?.toDouble() ?? 0.0,
      seatingCapacity: (json['seatingCapacity'] as num?)?.toInt() ?? 0,
      contactName: json['contactName']?.toString() ?? '',
      contactPhone: json['contactPhone']?.toString() ?? '',
      coverImage: json['coverImage']?.toString() ?? '',
      gallery: (json['gallery'] as List? ?? []).map((e) => e.toString()).toList(),
      menuImages: (json['menuImages'] as List? ?? []).map((e) => e.toString()).toList(),
      rejectionReason: json['rejectionReason']?.toString() ?? '',
      isOnline: json['isOnline'] != false,
      autoConfirm: json['autoConfirm'] == true,
    );
  }

  bool get isApproved => status == 'approved';
  bool get isPending => status == 'pending';
  bool get isRejected => status == 'rejected';
  bool get isSuspended => status == 'suspended';
}
